utils/file.go

```diff
@@ -29,3 +32,40 @@ func FileExists(filename string) bool {
 	_, err := os.Stat(filename)
 	return err == nil
 }
+
+// ShebangLine extracts the shebang line from the file.
+func ShebangLine(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	sc := bufio.NewScanner(f)
+	if !sc.Scan() {
+		// If the scan ended because of EOF, the file is empty, so sc.Err = nil.
+		// Otherwise sc.Err reflects the error reading the file.
+		return "", sc.Err()
+	}
+	line := sc.Text()
+	if !strings.HasPrefix(line, "#!") {
+		// Not a shebang line
+		return "", nil
+	}
+	return line, nil
+}
+
+// IsPOSIXShell attempts to detect POSIX-compliant shells (e.g bash, sh, zsh)
+func IsPOSIXShell(shell []string) bool {
+	bin := filepath.Base(shell[0])
+
+	if filepath.Base(shell[0]) == "env" {
+		bin = filepath.Base(shell[1])
+	}
```

Comment:
are there other shebang-isms that we should be checking for here? `#!/path/to/shell` and `#!/usr/bin/env shell` are the only two major ones i can think of, and are both handled here, but  are there others?

Reply:
It's possible on Linux to chain interpreters by making the shebang line refer to another script (with its own shebang line, which could be another script, which ...). I don't know who does this, and I'm inclined to avoid parsing arbitrary chains of scripts unless we come across a use-case.

It's possible for binaries to be in odd places (e.g. I think maybe Solaris puts bash somewhere unusual?) but `filepath.Base` handles that.

It's possible for the interpreter path to be a symlink to one of the supported shell binaries, but with a different name (e.g. `/usr/local/bin/frobulator -> /bin/bash`). The real-world use-cases I know of are symlinking `/bin/sh` to Bash or something else, but that would be handled already. I think following symlinks to resolve the name would be a reasonable addition to this function, but also would be as far as we should go.